File: UdemyCoursePy/src/mx/irving/arryas/FindMissingElement.py
import unittest
import collections


def find(array1, array2):
    difference = 0
    for index in range(len(array2)):
        difference += array1[index] - array2[index]
    return difference + array1[len(array1) - 1]


def finder(arr1, arr2):
    arr1.sort()
    arr2.sort()

    for num1, num2 in zip(arr1, arr2):
        if num1 != num2:
            return num1

    return arr1[-1]


def finder2(arr1, arr2):
    d = collections.defaultdict(int)

    # Add a count for every instance in Array 2
    for num in arr2:
        d[num] += 1

    # Check if num not in dictionary
    for num in arr1:
        if d[num] == 0:
            return num

        else:
            d[num] -= 1


def finder3(arr1, arr2):
    result = 0

    # Perform an XOR between the numbers in the arrays
    for num in arr1 + arr2:
        result ^= num

    return result


class TestFinder(unittest.TestCase):

    def test(self, sol):
        self.assertEquals(sol([5, 5, 7, 7], [5, 7, 7]), 5)
        self.assertEquals(sol([1, 2, 3, 4, 5, 6, 7], [3, 7, 2, 1, 4, 6]), 5)
        self.assertEquals(sol([9, 8, 7, 6, 5, 4, 3, 2, 1], [9, 8, 7, 5, 4, 3, 2, 1]), 6)
        print('ALL TEST CASES PASSED')


t = TestFinder()
t.test(find)
t.test(finder)
t.test(finder2)
t.test(finder3)

File: UdemyCourseKt/src/main/kotlin/mx/irving/arrays/FindMissingElement.kt
package mx.irving.arrays

fun find(array1: Array<Int>, array2: Array<Int>): Int {
    var difference = 0
    for (index in 0 until array2.size) {
        difference += array1[index] - array2[index]
    }
    return difference + array1.last()
}

fun finder(arr1: Array<Int>, arr2: Array<Int>): Int {
    arr1.sort()
    arr2.sort()

    for ((num1, num2) in arr1.zip(arr2)) {
        if (num1 != num2)
            return num1
    }
    return arr1.last()
}

fun finder2(array1: Array<Int>, array2: Array<Int>): Int {
    val map = hashMapOf<Int, Int>()

    for (num in array2) {
        map[num] = map[num]?.plus(1) ?: 1
    }

    for (num in array1) {
        if (map[num] == 0 || !map.containsKey(num))
            return num
        else map[num] = map[num]?.minus(1) ?: -1
    }
    return -1
}

fun finder3(array1: Array<Int>, array2: Array<Int>): Int {
    var result = 0
    // Perform an XOR operation between the numbers in the arrays
    for (num in array1 + array2) {
        result = result xor num
    }
    return result
}